Comments: Detail view

Copons opened the pull request:
Closes #13894 

This is the first code iteration of the **`CommentDetail`** block, part of the **Comment Management** effort.
Both the data structure and the design are currently temporary, but this block will be helpful for the Comment Management development, while also providing a sandbox to quickly try out design changes and improvements.

The `CommentDetail` block uses a flat props structure (as demoed in `blocks/comment-detail/example.jsx`) to be as ready as possible to changes in the actual data structure.
It also already features a bare minimum collection of internal state methods that might be eventually replaced by Redux actions.

The `CommentDetail` block is currently testable at: `/devdocs/blocks/comment-detail`.

![may-16-2017 19-34-50](https://cloud.githubusercontent.com/assets/2070010/26122307/eeefcfe2-3a6e-11e7-9956-7ceb2db28684.gif)

a8ci18n commented on client/blocks/comment-detail/comment-author-meta.jsx:
```diff
@@ -0,0 +1,34 @@
+		<div>
+			<Gridicon icon="comment" size={ 18 } />
+			{ translate(
+				'%d Comment', '%d Comments',
```
Hi! I've found a possible matching string that has already been translated **37** times:
`translate( '%d comment', '%d comments' )` *ES Score: 20*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/comments/detail-view)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -0,0 +1,112 @@
+					<div className="comment-detail__author-more-actions">
+						<a className="comment-detail__author-more-element">
+							<Gridicon icon="comment" />
+							<span>{ translate( 'View all comments' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'View All Comments', { context: 'NewDash Comments Activity Feed'} )` *ES Score: 10*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/comments/detail-view)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

gwwar replied:
As noted in the hangout we're punting the user comment view to the second milestone.

gwwar replied:
See https://github.com/Automattic/wp-calypso/issues/13895

It's fine to work on this if it turns out to be easier than we expect, but let's leave wiring pf author info for last, since we can cut this out of m1 and have a useable product.

Copons replied:
Ah yeah, I went all autopilot here and totally forgot about that.
I'll remove that button, it's easy to add it back in anyway.

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -0,0 +1,112 @@
+			<div className={ classes }>
+				<div className="comment-detail__author-preview">
+					<div className="comment-detail__author-avatar">
+						<img src={ author.avatar_URL } />
```
As we chatted about earlier, let's use flat props in our new components.  For example `authorAvatarURL` vs `author.avatar_URL`

This does two things:
1. Guards against future data shape changes
2. Makes the data needs of the component much more explicit

Copons replied:
Oh, about this: I think the moment I started dropping the flat props in favour of nested objects was actually when I've introduced `ReaderAvatar`, that has an `author` prop instead of 4 different flat props.
See the mock author in its doc file: https://github.com/Automattic/wp-calypso/blob/master/client/blocks/reader-avatar/docs/example.jsx#L12-L17

I've taken the same mock author, and in no time I've found myself nesting everything into everything.

gwwar replied:
Right and we used to have a pattern where we passed though a full `site` object all over 😄 . 
 
User is a much smaller object, so I'd prefer if we used flat props, but I wouldn't block this PR. (That 🐍  case though. Hah.)

On that note, as a heads up `users` hasn't been fully ported to the redux store. We currently are only guaranteed that the current-user is available:  https://github.com/Automattic/wp-calypso/blob/master/client/boot/common.js#L179

mtias replied:
I suggest not using ReaderAvatar in the context of My Sites. We favor site icon as the visual ID, and this makes it very hard to quickly identify things.

gwwar replied:
@mtias in wp-admin, it looks like we're using the user avatar over a site icon. cc @drw158

mtias replied:
Yep :)

gwwar commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -0,0 +1,112 @@
+		} = this.props;
+		const { isExpanded } = this.state;
+
+		const classes = classNames( 'comment-detail__author', {
```
Do we get a warning for this? I would think the class name would be `comment-detail-author` because it's the full component name.

Copons replied:
It would be a warning if the component was in an index file in its own folder, e.g. `comment-detail/comment-detail-author/index.jsx`.

In this case, it's considered a "sub-component" of `comment-detail`.

gwwar commented on client/blocks/comment-detail/comment-detail-comment.jsx:
```diff
@@ -0,0 +1,48 @@
+					<div className="comment-detail__comment-reply">
+						<a>
+							<Gridicon icon="reply" />
+							<span>{ translate( 'You replied to this comment' ) }</span>
```
@drw158 does this expand the author's response if we click on this?

gwwar commented on client/blocks/comment-detail/index.jsx:
```diff
@@ -0,0 +1,67 @@
+					isExpanded={ isExpanded }
+					toggleExpanded={ this.toggleExpanded }
+				/>
+				{ isExpanded &&
```
Was [FoldableCard](https://wpcalypso.wordpress.com/devdocs/design/foldable-card) not a good fit?

Copons replied:
I've tried using FoldableCard at first, but then I felt there was just too much overriding to justify it.
Also I kept the repetition to a minimum: pretty much only the margin transition and the internal state for toggling the expanded view.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,261 @@
+
+.comment-detail__author-avatar {
+	flex-shrink: 0;
+	img {
```
Please avoid any bare element selectors, we can add a class to the img instead.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,261 @@
+	}
+}
+.comment-detail__author-info {
+	@extend %comment-detail__ellipsis;
```
I suppose we do suggest using sass `extend` in our css guidelines. While `extend` does make the unminified file smaller, usually you’re usually better off just repeating yourself since the gzipped file will likely be smaller. (gzip is great at compressing repeated strings).

Note that we currently can't use extend across media queries, so if we think we'll be adding any, we should avoid this. https://github.com/sass/sass/issues/1050

cc @mtias if you had other thoughts

Copons replied:
Yeah, it was mostly for me to avoid forgetting some rules.

Maybe I'm better off using an actual class instead?

gwwar replied:
So if we find that we're repeating ourselves, we normally try to break things out as a components. Which would work well to change `%comment-detail__gray-link;` into a new comment link component or as a composed link inside of a comment compoennt.

`%comment-detail__ellipsis` is a bit tougher. A few options here:
1. keep the extends (though we'd need to convert to a mixin if we add media queries). See: https://github.com/Automattic/wp-calypso/blob/master/assets/stylesheets/shared/mixins/_mixins.scss#L17-L31
2. just repeat the rules in the related classes

Copons replied:
Mmh, I'll probably drop both altogether actually.

`%comment-detail__ellipsis` will be an ugly repetition, but eh, I'm kinda confident we won't have to use it many more times.

`%comment-detail__gray-link` this actually is a sort of placeholder.
Like, in this PR the links are all gray and blue on hover (like in the Reader's comments).
But, for example, in the notification panel, on hover the icon remains gray and the label becomes yellow.
Also the various actions will most likely have "states" (e.g. approved might turn green, spam red), leading to an inevitable overhaul once the design is final.

gwwar replied:
> Also the various actions will most likely have "states" (e.g. approved might turn green, spam red), leading to an inevitable overhaul once the design is final.

A modifier class might be a good fit in that case. eg `is-spam`, `is-approved`

Copons replied:
> A modifier class might be a good fit in that case. eg is-spam, is-approved

Ah yes of course, but it'll anyway require some proper thinking instead of my blanket "links are gray and let's call it a day" approach. 😄

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,261 @@
+	border-bottom: 1px solid lighten( $gray, 30% );
+	padding-left: 16px;
+
+	.comment-detail__comment-content {
```
Did we need the extra CSS specificity here? Otherwise this doesn't need to be nested.

gwwar replied:
See also https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#nesting

Copons replied:
We rarely do in this stylesheet.
It's just to improve code readability

gwwar replied:
Ah so as a general rule, we make our classnames so long so we can just target them directly. This leaves CSS specificity at (0, 1, 0) (ids, classes, elements), and makes it easy to override rules (which is often a problem in large application css files). Thee are exceptions of course, but mainly we only do so for component composition or modifier classes.

See also: https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md#csssass-coding-guidelines  
> Since every element will have a direct class to be selected with, we avoid descendant selectors and child selectors. We also avoid indenting selectors in Sass files in favor of single selectors. The main exceptions are precisely the is-modifier class, and the cases where the context needs to change a given component's accidents.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,261 @@
+	}
+}
+
+@keyframes author-more-info {
```
Maybe namespace this with `comment-detail`

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,261 @@
+
+	.comment-detail__author-info {
+		@extend %comment-detail__ellipsis;
+		> div {
```
Can we add a classname to this div? We'd be able to avoid  the `>` and the deep nesting here

gwwar commented on client/blocks/comment-detail/comment-detail-post.jsx:
```diff
@@ -0,0 +1,34 @@
+	url,
+} ) =>
+	<div className="comment-detail__post">
+		<ReaderAvatar
```
:D I wonder if we should rename the `ReaderAvatar` to something more generic in a future PR if we end up using this outside of the reader. cc @blowery

Copons replied:
Wait for that tho: this was actually my own addition, the current design iteration doesn't use any avatar there. 🙂 
I've just stumbled upon that component while browsing the design docs and thought it would look good here.

gwwar replied:
> I've just stumbled upon that component while browsing the design docs and thought it would look good here.

Me too! When I checked I was happy that we were already using it.

gwwar commented:
Overall, this looks like a fantastic start! Great work @Copons. Feel free to address some of the issues, and when you're happy with it let's 🚢 

We can continue to iterate in another PR if there's any comments that are left.

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -0,0 +1,132 @@
+							<Gridicon icon="block" />
+							<span>{
+								authorIsBlocked
+									? translate( 'Unblock user' )
```
Hi! I've found a possible matching string that has already been translated **39** times:
`translate( 'Unblock' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Copons commented:
@gwwar to recap all the changes:

- Converted all object props to flat props - with the side effect of doubling the LOC, but I assume that once connected to Redux, most of the props inheritance will be removed.
- Replaced `ReaderAvatar` with a simple site icon image: design iterations still don't use any image there, so I'd say don't clutter the props even more than what's strictly needed.
- Removed most element selectors - not all of them, in some cases (such as most `a` and some `span` used as quick style modifier) I think it doesn't make much sense.
- Replaced `@extend` with simple repetitions.
- Removed unnecessary style nesting.
- Better style namespacing.
- Added actual functions for the various state changes (e.g. `toggleApprove()`); right now they only change the icon colors, but can be easily modified to perform some real actions.

Due to the amount of changes on the props inheritance, I'd appreciate a last quick look before actually shipping, to be sure I got it right. 🙂

gwwar commented on client/blocks/comment-detail/comment-detail-post.jsx:
```diff
@@ -0,0 +1,31 @@
+	siteName,
+} ) =>
+	<div className="comment-detail__post">
+		<div className="comment-detail__site-icon">
```
Could we try using the SiteIcon block?

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,322 @@
+.comment-detail__reply {
+	line-height: 0;
+
+	textarea {
```
We can give this a classname later. It's unlikely we'd have multiple text areas.

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,322 @@
+	}
+}
+
+.comment-detail.author-is-blocked .comment-detail__author-more-element-block-user .gridicon {
```
I think we can simplify this a bit, as the specificity here is quite high. For example if we gave the gridicon a classname, and moved the modifer closer:
```jsx
.comment-detail__block-icon.is-blocked {
    color: $alert-red;
    fill: $alert-red;
}
```

gwwar commented on client/blocks/comment-detail/style.scss:
```diff
@@ -0,0 +1,322 @@
+	transform: scaleY( 0 );
+	transform-origin: top;
+}
+.comment-detail__author.is-expanded .comment-detail__author-more-info {
```
Unless things are particularly awkward to apply in bulk, I would recommend keeping the modifier as close to the actual element we're toggling since it keeps the rule short and the CSS specificity low.

It's optional here, but keep it in mind when we create very generic components, as it can be a pain to override rules when we get too specific.

gwwar commented:
@Copons I took the branch for another spin, and things look good. I left a few notes, but nothing that blocks this.

Do you mind updating the summary to provide a bit more context before 🚢

Copons commented:
@gwwar My reasoning behind placing modifier classes as early as possible is exactly to give this component as much freedom as possible during the future design iterations.

Think for example of `.is-approved`.
In this draft it only toggles the approved icon active state, so it makes sense to put it as close as possible to the icon itself.
Though, we might want to change the color of both icon and label - and that's why after your suggestion I've decided to add it to the link: `.comment-detail__action-approve.is-approved`.
At the same time, we might want to change the color of the entire comment when it's unapproved, so I've also left the class on the root element: `.comment-detail.is-approved`.

Same thing for `.is-expanded`: at some point we might want to, say, show more stuff on comment expansion, or do some animation and whatnot.
Having the modifier class on top of everything allows us to do whatever we want without the burden of moving props around.

---

This said, I've introduced `SiteIcon` - which I totally missed before!

I'll merge as soon as I update the PR summary. Thanks again for the reviews! 🙇

a8ci18n commented on client/blocks/comment-detail/comment-detail-author.jsx:
```diff
@@ -0,0 +1,135 @@
+							<Gridicon icon="block" />
+							<span>{
+								authorIsBlocked
+									? translate( 'Unblock user' )
```
Hi! I've found a possible matching string that has already been translated **39** times:
`translate( 'Unblock' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).
